LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultLoRaDataRequestHandler.cs

```diff
@@ -64,514 +65,541 @@ public class DefaultLoRaDataRequestHandler : ILoRaDataRequestHandler
 
         public async Task<LoRaDeviceRequestProcessResult> ProcessRequestAsync(LoRaRequest request, LoRaDevice loRaDevice)
         {
-            if (request is null) throw new ArgumentNullException(nameof(request));
-            if (loRaDevice is null) throw new ArgumentNullException(nameof(loRaDevice));
+            ArgumentNullException.ThrowIfNull(request, nameof(request));
+            ArgumentNullException.ThrowIfNull(loRaDevice, nameof(loRaDevice));
 
             List<Task> deferredTasks = null;
+            IAsyncDisposable deviceConnectionActivity = null;
 
-            var timeWatcher = request.GetTimeWatcher();
-
-            var loraPayload = (LoRaPayloadData)request.Payload;
-            this.d2cPayloadSizeHistogram?.Record(loraPayload.Frmpayload.Length);
-
-            var payloadFcnt = loraPayload.Fcnt;
+            try
+            {
+                ExceptionDispatchInfo edi = null;
+                LoRaDeviceRequestProcessResult result = null;
 
-            var payloadFcntAdjusted = LoRaPayloadData.InferUpper32BitsForClientFcnt(payloadFcnt, loRaDevice.FCntUp);
-            this.logger.LogDebug($"converted 16bit FCnt {payloadFcnt} to 32bit FCnt {payloadFcntAdjusted}");
+                try
+                {
+                    result = await ProcessAsync();
+                }
+#pragma warning disable CA1031 // Do not catch general exception types (exception is rethrown)
+                catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+                {
+                    edi = ExceptionDispatchInfo.Capture(ex);
+                }
 
-            var requiresConfirmation = loraPayload.RequiresConfirmation;
+                Exception deferredTaskException = null;
+                if (deferredTasks is { } someDeferredTasks)
+                {
+                    _ = await Task.WhenAny(Task.WhenAll(someDeferredTasks));
+                    var exs = someDeferredTasks.GetExceptions();
 
-            LoRaADRResult loRaADRResult = null;
+                    deferredTaskException = exs switch
+                    {
+                        { Count: 1 } => exs[0],
+                        { Count: > 1 } => new AggregateException(exs),
+                        _ => null
+                    };
+                }
 
-            var frameCounterStrategy = this.frameCounterUpdateStrategyProvider.GetStrategy(loRaDevice.GatewayID);
-            if (frameCounterStrategy == null)
-            {
-                this.logger.LogError($"failed to resolve frame count update strategy, device gateway: {loRaDevice.GatewayID}, message ignored");
-                return new LoRaDeviceRequestProcessResult(loRaDevice, request, LoRaDeviceRequestFailedReason.ApplicationError);
-            }
+                if (edi is { } someEdi)
+                {
+                    if (deferredTaskException is { } someDeferredTaskException)
```

Comment:
Doesn't having this nested here mean that no errors will be logged if one or more of the forked tasks failed unless the main processing also failed? Shouldn't we be logging irrespective? That is, as long as `deferredTaskException` isn't `null`, which was established earlier?

Reply:
This was simplified with 8a82a1f3a5182e059c5e29f4cc08c7495d6eb485